### PR TITLE
Fix protractor retries and screenshots

### PR DESCRIPTION
**Pull request by garethbowen**
# Description

medic/cht-core#8103

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8103-protractor-retries/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8103-protractor-retries/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8103-protractor-retries/docker-compose/cht-couchdb-clustered.yml)
 
# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/utils.js [hunk 1]
```diff
@@ -1338,7 +1340,7 @@ module.exports = {
   tearDownServices: stopServices,
   endSession: async (exitCode) => {
     await module.exports.tearDownServices();
-    return module.exports.reporter.afterLaunch(exitCode);
+    await new Promise((resolve) => module.exports.reporter.afterLaunch(resolve.bind(this, exitCode)));
```
Comment (garethbowen) [inline tests/utils.js:1343]: This was the root issue, predating but perhaps exacerbated by, the retries change - the screenshot reporter afterLaunch function expects a callback to be passed.

**Review by dianabarsan — COMMENTED** [review]
Looking at the failure captured here: https://github.com/medic/cht-core/issues/8103#issuecomment-1445553380, it seems like only the _last_ retry snapshot is kept, which makes me think that all the previous snapshots were still deleted. 
I also gave it a try, intentionally failing one e2e test, and got the same. 

I know we don't intend to keep protractor around for long, so I don't think this is worth a lot of effort. 
I think a lot of test retries will fail because no cleanup is made after previous test (like in the case of my test: https://github.com/medic/cht-core/actions/runs/4282260079), so the last screenshot that is saved might have little value compared to the first. 
Is it possible to make sure we don't clean up the results folder at all?

**Comment by garethbowen** [comment]
Ooof. I've spent some time looking into this.

So... retries (especially in protractor) didn't seem to solve any real flaky tests because the setup and teardown didn't happen between retries in a way that would allow the test to re-run. That, combined with the fact we're dropping protractor anyway, makes me reluctant to spend too much time on it - better to keep improving WDIO instead.

If you agree, I'll revert all the retry code in Protractor to get us back to where we were before, with the same level of flake but screenshots so we can debug it.

@dianabarsan What do you think?

**Comment by dianabarsan** [comment]
> If you agree, I'll revert all the retry code in Protractor to get us back to where we were before, with the same level of flake but screenshots so we can debug it.

Yes, please!

**Comment by garethbowen** [comment]
@dianabarsan Retries removal review please!

**Review by dianabarsan — APPROVED** [review]
Thanks for looking into this! I'm sorry we only ended up back where it started.

**Pull request merged**